### PR TITLE
Add image processing example

### PR DESCRIPTION
**Pull request by gbryant-arm**
Adding image processing to the examples.
~~It must be built with the nightly toolchain, as specified in the toolchain file. This is due to `proper`, a dependency of `image`, using an unstable feature.~~ It doesn't have to be built with the nightly toolchain anymore :man_shrugging: 

The data generator fetches images from the web, and has the following benefits:
* no need to hardcode images in the git index
* copyright compliant
* access to realistic pictures, as opposed to computer generated ones
* code can be reused to fetch any kind of test data

Execution:
* Go to `sdk/freestanding-execution-engine`
* Run the execution engine with `test.png` mapped to the vfs:
  `cargo run -- -d test.png -p image-processing.wasm -x interp`

### CODE REVIEW

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: d1599930a120d18c930481ad0a2b923a12a3b2c0
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=e444b9f0-7a53-4dc2-9c3d-0e54224b2cdd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: sdk/rust-language-support/examples/image-processing/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+//! See the `LICENSE.markdown` file in the Veracruz root directory for
+//! information on licensing and copyright.
+
+extern crate image;
+
+use image::{GenericImageView, imageops, ImageFormat, io::Reader};
+use std::process::exit;
+use wasi_types::ErrNo;
+
+/// Reads two input sources from the Veracruz host: the first input source is assumed to be a vector
```
Comment (dominic-mulligan-arm) [inline sdk/rust-language-support/examples/image-processing/src/main.rs:34]: This `rustdoc` comment is incorrect.  It needs replacing with something specific to this code.

File: sdk/rust-language-support/examples/image-processing/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+/// structs.  Fails with [`return_code::ErrorCode::BadInput`] if the number of inputs provided is
+/// not equal to 2, or if the inputs cannot be deserialized from Bincode.
+/// TODO: read image directly (stream)
+
+fn process_image() -> Result<(), wasi_types::ErrNo> {
+    // Use the open function to load an image from a Path.
+    // `open` returns a `DynamicImage` on success.
+    // By default, jpeg-decode reads JPEG images in parallel threads (cf. rayon), which aren't supported in WASM. The workaround is to read PNG images instead
+    let mut img = image::open("/test.png").map_err(|_| {
+        println!("Failure opening /test.png");
```
Comment (dominic-mulligan-arm) [inline sdk/rust-language-support/examples/image-processing/src/main.rs:45]: What will `println!` do, here, when executed?

File: sdk/rust-language-support/examples/image-processing/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        ErrNo::Proto
+    })?;
+
+    // Transform the image
+    let subimg = imageops::crop(&mut img, 0, 0, 100, 100);
+    println!("new dimensions: {:?}", subimg.dimensions());
+
+    // Write the contents of this image to the Writer in PNG format.
+    subimg.to_image().save_with_format("/output", ImageFormat::Png).unwrap();
```
Comment (dominic-mulligan-arm) [inline sdk/rust-language-support/examples/image-processing/src/main.rs:54]: Do we want an `unwrap`, and therefore a runtime panic, or some more graceful error notification, here?

File: sdk/rust-language-support/examples/image-processing/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    // Verify the output
```
Comment (dominic-mulligan-arm) [inline sdk/rust-language-support/examples/image-processing/src/main.rs:56]: Why is there commented code left in a pull request?

**Review by dominic-mulligan-arm — CHANGES_REQUESTED** [review]

**Comment by dominic-mulligan-arm** [comment]
Is this going to get built when the main Makefile builds `sdk`?

File: sdk/rust-language-support/examples/image-processing/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+image = "0.23.14"
+wasi-types = { git = "https://github.com/veracruz-project/wasi-types.git", branch = "veracruz" }
+
+[profile.release]
+codegen-units = 1
+opt-level = 3
```
[the remark this answers is not in the document]
Comment (gbryant-arm) [reply sdk/rust-language-support/examples/image-processing/Cargo.toml:16]: Thanks for the spotting. Any reason why some examples don't have `lto` enabled?

**Comment by gbryant-arm** [comment]
Ready for review

File: sdk/rust-language-support/examples/image-processing/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (dominic-mulligan-arm) [reply sdk/rust-language-support/examples/image-processing/Cargo.toml:17]: From memory, I'm not sure.  I think at one point link-time optimisation may have been broken, but can't say for sure.

File: sdk/rust-language-support/examples/image-processing/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+//! See the `LICENSE.markdown` file in the Veracruz root directory for
+//! information on licensing and copyright.
+
+extern crate image;
+
+use image::{GenericImageView, imageops, ImageFormat};
+use std::process::exit;
+use wasi_types::ErrNo;
+
+fn process_image() -> Result<(), wasi_types::ErrNo> {
```
Comment (dominic-mulligan-arm) [inline sdk/rust-language-support/examples/image-processing/src/main.rs:30]: Can you add top-level `rustdoc` comments to this function, explaining what it does?  I think the comment about `jpeg` decoding and Rayon threads can be moved into the `rustdoc` comment, too.

**Review by dominic-mulligan-arm — CHANGES_REQUESTED** [review]

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: b4e047d98a2a2dbd2df09f15aba43551cc3bf309
* Result: FAILED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=778b7feb-486d-49e0-9aba-e8f0dd87057b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: df179bfefb4c860c91d98479e922808b084db3da
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=bff020cf-95ed-44f0-8e7c-2aa69f2ee450%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 7eaa243b60fe88a2ce74353590efad4b757bc9ce
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=f3cd6c32-54b2-4c58-a28b-636de72cf572%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 4e492bda36b9c60c86069e1021e28559504ba2d3
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=ffd01a70-bd17-4dfb-bcd6-17d492d9b9cb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 5c1bfbd3df42bfbcb54ef68962513a469f48e353
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=c47a74d2-9f42-41f6-a07d-4ef374201ed5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 17082df4994e1fc2ba467002bc91eea74b6f003e
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=0987266c-becd-4b8e-b9f7-d01e012bde64%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: e1ee83e16d294d88e437d6d49e386fc0e22b8eac
* Result: FAILED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=fea2259f-0b6c-4160-84ee-1bb01dcd4fb2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by dominic-mulligan-arm — DISMISSED** [review]

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 30b2e215a686a92cf1f3c38e8c640bef1958de18
* Result: FAILED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=18ff283a-f709-4ce4-b764-335cbde49038%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: sdk/rust-examples/image-processing/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    // Use the open function to load an image from a Path.
+    // `open` returns a `DynamicImage` on success.
+    let mut img = image::open("/test.png").map_err(|_| ErrNo::Io)?;
+
+    // Transform the image
+    let subimg = imageops::crop(&mut img, 0, 0, 100, 100);
+    println!("new dimensions: {:?}", subimg.dimensions());
+
+    // Write the contents of this image to the Writer in PNG format.
+    subimg.to_image().save_with_format("/output", ImageFormat::Png).map_err(|_| ErrNo::Io)?;
```
[the remark this answers is not in the document]
Comment (gbryant-arm) [reply sdk/rust-examples/image-processing/src/main.rs:42]: Thanks @geky!

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 227de7ee8338dff04230dfc7ffc974c2b4930711
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=880d4b8f-b343-4c6f-91cf-fe0362c1809f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: a413f1a4eb566da6ac3fda4b6c8fc9210674f168
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=55f29d8d-6bc0-4e72-8fa7-e392ffc44e5a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: aaba5aeefce46fab8e5c74dba5645404dc8d18a2
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=268573c7-9fc8-4d91-8440-699092983d68%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by dominic-mulligan-arm** [comment]
@gbryant-arm can:

1. This be rebased onto `main`,
2. Be rewritten to follow the pattern that @geky used in #158 with the `anyhow::Result<_>` type?

Moreover, what is the plan for test data for this example?  For our other examples, we have "generating" programs that produce synthetic test data which allows us to perform tests without necessarily having to include data in the repository, neatly avoiding issues with licensing and attribution.

**Comment by gbryant-arm** [comment]
> Moreover, what is the plan for test data for this example? For our other examples, we have "generating" programs that produce synthetic test data which allows us to perform tests without necessarily having to include data in the repository, neatly avoiding issues with licensing and attribution.
I see. I can have a look into generating images

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 06c6a6194454a6c5023ad5ab02c4f2e8d823f457
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=a508782e-b200-463f-ba65-d90888bad288%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 9ae389465069f27f97dfb09997dedd91c70bc313
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=75226314-b2f0-41ee-b486-0ae16f1899fc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 23dcc2a839f540c3ff5858e5eca51141cbc88c46
* Result: FAILED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=d414e8b3-db5c-4db5-8ee2-c6b052575618%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 990002bb14372843b0e326bdef4d0fb3a3c1b887
* Result: FAILED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=0f3ce1e8-13e8-4bf2-8c44-16a4a8491d2c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: e8f3c4107919af63fa6459c1c2acdd0be78f452e
* Result: SUCCEEDED
* [Build Logs](https://t4r5bdzqe9.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=4cfe4577-5716-4cb7-92ea-026a2614a844%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by ShaleXIONG — APPROVED** [review]

**Comment by dominic-mulligan-arm** [comment]
Can you rebase this, @gbryant-arm?

File: sdk/data-generators/image-processing-generator/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+/// Parameters:
+/// * `file_prefix`, String, the prefix of the generated files.
+/// * `width`, u64, the image width, default is 10.
+/// * `height`, u64, the image height, default is 10.
+fn main() -> Result<(), Box<dyn Error>> {
+    let matches = App::new("Data generator for image processing")
+        .version("pre-alpha")
+        .author("The Veracruz Development Team")
+        .about("Download a random JPEG image of dimensions ([WIDTH], [HEIGHT]) and convert it to a PNG image") 
+       .arg(
```
Comment (dominic-mulligan-arm) [inline sdk/data-generators/image-processing-generator/src/main.rs:33]: Can you run `cargo fmt` on the file?  There's some bad indentation here...

File: sdk/data-generators/image-processing-generator/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    let image_height = matches
+        .value_of("height")
+        .ok_or("Failed to read the height")?;
+
+    // Download random image
+    let mut downloader = Downloader::builder()
+        .download_folder(std::path::Path::new("./"))
+        .parallel_requests(1)
+        .build()
+        .unwrap();
```
Comment (dominic-mulligan-arm) [inline sdk/data-generators/image-processing-generator/src/main.rs:89]: Can you replace this `unwrap()` with an `expect()` and a meaningful message indicating what went wrong?

**Comment by dominic-mulligan-arm** [comment]
@gbryant-arm have you finished addressing review comments for this change?  If so, can you rebase in preparation for final review and merging?

**Comment by gbryant-arm** [comment]
I addressed all the comment above. Rebasing now

**Review by dominic-mulligan-arm — APPROVED** [review]

**Comment by dominic-mulligan-arm** [comment]
+1 reached, merging.

**Pull request merged**